torchmultimodal/modules/layers/attention.py

```diff
@@ -0,0 +1,240 @@
+        self.w_qs.weight.data.normal_(std=1.0 / torch.sqrt(torch.tensor(dim_q)))
+
+        self.w_ks = nn.Linear(dim_kv, n_head * self.d_k, bias=False)  # k
+        self.w_ks.weight.data.normal_(std=1.0 / torch.sqrt(torch.tensor(dim_kv)))
+
+        self.w_vs = nn.Linear(dim_kv, n_head * self.d_v, bias=False)  # v
+        self.w_vs.weight.data.normal_(std=1.0 / torch.sqrt(torch.tensor(dim_kv)))
+
+        self.fc = nn.Linear(n_head * self.d_v, dim_q, bias=True)  # c
+        self.fc.weight.data.normal_(std=1.0 / torch.sqrt(torch.tensor(dim_q * n_layer)))
+
+        self.attn = attn_module
+
+        self.cache: Dict[str, Tensor] = dict()
+
+    def _split_multihead(self, x: Tensor) -> Tensor:
+        # Splits input tensor of size (b x (d1...dn) x hidden)
+        # into (b x (d1...dn) x n_head x emb_dim)
+        x = x.unflatten(-1, (self.n_head, -1))
+        # Rearrange to put head dim first, (b x n_head x (d1...dn) x emb_dim)
+        x = shift_dim(x, -2, 1)
+        return x
+
+    def _combine_multihead(self, x: Tensor) -> Tensor:
+        # Moves head dim back to original location and concatenates heads
+        # (b x n_head x (d1...dn) x emb_dim) -> (b x (d1...dn) x hidden)
+        return shift_dim(x, 1, -2).flatten(start_dim=-2)
+
+    def forward(
+        self, q: Tensor, k: Tensor, v: Tensor, decode_step=None, decode_idx=None
+    ) -> Tensor:
+        # compute k, q, v
+        d_k, d_v, n_head = self.d_k, self.d_v, self.n_head
+        q = self._split_multihead(self.w_qs(q))
+        k = self._split_multihead(self.w_ks(k))
+        v = self._split_multihead(self.w_vs(v))
+
+        # fast decoding
+        if decode_step is not None:
+            if decode_step == 0:
+                if self.causal:
+                    k_shape = (
+                        q.shape[0],
+                        n_head,
+                        *self.shape,
+                        self.d_k,
+                    )  # B x n_head x 3 x 32 x 32 x d
+                    v_shape = (q.shape[0], n_head, *self.shape, self.d_v)
+                    self.cache = dict(
+                        k=torch.zeros(k_shape, dtype=k.dtype, device=q.device),
+                        v=torch.zeros(v_shape, dtype=v.dtype, device=q.device),
+                    )
+                else:
+                    # cache only once in the non-causal case
+                    self.cache = dict(k=k.clone(), v=v.clone())
+            if self.causal:
+                idx = (
+                    slice(None, None),
+                    slice(None, None),
+                    *[slice(i, i + 1) for i in decode_idx],
+                )
+                self.cache["k"][idx] = k
+                self.cache["v"][idx] = v
+            k, v = self.cache["k"], self.cache["v"]
+
+        a = self.attn(q, k, v, decode_step, decode_idx)
+        a = self._combine_multihead(a)
+        a = self.fc(a)
+
+        return a
+
+
+class FullAttention(nn.Module):
```

Comment:
is the above MHA + Full Attention the same as the MHA from core? if yes, why not use that for gpt and have an implementation for axial attention?

Reply:
that would still require our own MHA class. was also thinking we could punt consolidating with core later

Reply:
resolved offline